google-http-client/src/main/java/com/google/api/client/http/apache/ApacheHttpRequest.java

```diff
@@ -60,6 +70,21 @@ public LowLevelHttpResponse execute() throws IOException {
       ((HttpEntityEnclosingRequest) request).setEntity(entity);
     }
     request.setConfig(requestConfig.build());
-    return new ApacheHttpResponse(request, httpClient.execute(request));
+    if (dnsResolver == null) {
+      // Use the default OS DNS resolution.
+      return new ApacheHttpResponse(request, httpClient.execute(request));
+    }
+
+    // Use the IP provided by our specialized DNS resolver.
+    InetAddress ip = dnsResolver.resolve(request.getURI().getHost());
+    int port =
+        request.getURI().getPort() != -1
+            ? request.getURI().getPort()
+            : "https".equals(request.getURI().getScheme()) ? 443 : 80;
```

Comment:
What happens if we provide a URI with weird casing? `URI.create("hTTPs://google.com").getURI().getScheme()` returns `"hTTPs"`. Maybe we don't care what happens in that case.